spec.bs

```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If |worklet|'s [=addModule initiated=] is true, return "DisallowedDueToNonPreferenceError".
     1. Set |worklet|'s [=addModule initiated=] to true.
-    1. Return true.
+    1. If |workletOrigin| does not equal [=this=]'s [=relevant settings object=]'s [=environment settings object/origin=], set |worklet|'s [=SharedStorageWorklet/is cross-origin worklet=] to true.
+    1. If |worklet|'s [=cross-origin worklet allowed=] is false, and if |worklet|'s [=SharedStorageWorklet/is cross-origin worklet=] is true, return "DisallowedDueToNonPreferenceError".
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given [=this=]'s [=relevant settings object=] and |workletOrigin| is false, return "DisallowedDueToPreferenceError".
+    1. Set |worklet|'s [=SharedStorageWorklet/worklet origin=] to |workletOrigin|.
+    1. Return "Allowed".
   </div>
 
 Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=], initially empty, can contain at most one instance of its [=worklet global scope type=], the {{SharedStorageWorkletGlobalScope}}.
 
+  Note: The website that serves the module script should be aware of the implication of CORS: when the the module script's URL's origin is cross-origin with the worklet's creator window's origin, and by granting the module script resource via CORS, it will also grant the worklet's creation and subsequent operations on the worklet, under module script's URL's origin. The worklet's creator context could poison and use up the [=SharedStorageWorklet/worklet origin=]'s budget.
+
+  ## Run Operation Methods on {{SharedStorageWorklet}} ## {#run-op-shared-storage-worklet}
+
+  <div algorithm>
+    To <dfn>get the select-url result index</dfn>, given {{SharedStorageWorklet}} |worklet|, {{DOMString}} |operationName|, [=/list=] |urlList|, and {{SharedStorageRunOperationMethodOptions}} |options|:
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |window| be [=current global object=].
+    1. Assert that |window| is a {{Window}}.
+    1. Let |context| be |window|'s [=Window/browsing context=].
+    1. If |context| is null, return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. If |environment|'s [=associated document=] is not [=fully active=], return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |realm| be the [=current realm=].
+    1. Let |outsideSettings| be |worklet|'s [=relevant settings object=].
+    1. Let |agent| be the result of [=obtaining a worklet agent=] given |outsideSettings|.
+    1. Run the following steps in |agent|:
+        1. Let |index| be [=default index=].
+        1. If |worklet|'s [=module map=] is not [=map/empty=]:
```

Comment:
I'm a little confused about the workflow here. We just barely created the _brand new_ global at this point, so how could its operation map be non-empty at this point? Don't we have to let the script run and let it register its functions? At least I hope we're not trying to store any Web IDL callback function objects in the operation map and reuse them among globals... that seems bad. Real Worklets in HTML seem to store the module script response in the response map, but then only register the "name" of each operation. Then when an operation is run, we use the realm-agnostic name string to pluck the actual function off of the specific global we've selected to execute within, and then run that realm-bound function to execute the operation in the worklet.

Reply:
[addModule](https://html.spec.whatwg.org/multipage/worklets.html#dom-worklet-addmodule)() (or createWorklet()) internally should invoke the script, which could invoke the [register()](file:///Users/yaoxia/Desktop/spec.html#dom-sharedstorageworkletglobalscope-register) method in the global scope.

This should be similar to [animation worklet’s registerAnimator](https://www.w3.org/TR/css-animation-worklet-1/#dom-animationworkletglobalscope-registeranimator).

Reply:
OK, and we know that this `addModule()` has always been called by this point because `selectURL()` throws a TypeError before this point if the list of global scopes is empty? That's the idea, right?

So in that case, I guess my question is now the opposite: how could the module map ever be _empty_ at this point? I'm assuming if we ever get this far, then we already have a worker module script that has loaded successfully, or is that wrong?

Reply:
Seems like this can be resolved given https://github.com/WICG/shared-storage/pull/131#issuecomment-1996109464.